src/java.base/share/classes/java/lang/reflect/code/descriptor/MethodDesc.java

```diff
@@ -69,38 +72,56 @@ static MethodDesc method(Method m) {
     }
 
     static MethodDesc method(Class<?> refType, String name, MethodType mt) {
-        return method(JavaType.type(refType), name, MethodTypeDesc.methodType(mt));
+        return method(refType, name, mt.returnType(), mt.parameterList());
     }
 
     static MethodDesc method(Class<?> refType, String name, Class<?> retType, Class<?>... params) {
-        return method(JavaType.type(refType), name, methodType(retType, params));
+        return method(refType, name, retType, List.of(params));
     }
 
     static MethodDesc method(Class<?> refType, String name, Class<?> retType, List<Class<?>> params) {
-        return method(JavaType.type(refType), name, methodType(retType, params));
+        return method(JavaType.type(refType), name, JavaType.type(retType), params.stream().map(JavaType::type).toList());
     }
 
-    static MethodDesc initMethod(MethodTypeDesc mt) {
-        return new MethodDescImpl(
-                mt.returnType(),
-                "<init>",
-                MethodTypeDesc.methodType(JavaType.VOID, mt.parameters()));
-    }
 
-    static MethodDesc method(TypeElement refType, String name, MethodTypeDesc type) {
+    static MethodDesc method(TypeElement refType, String name, FunctionType type) {
         return new MethodDescImpl(refType, name, type);
     }
 
     static MethodDesc method(TypeElement refType, String name, TypeElement retType, TypeElement... params) {
-        return method(refType, name, methodType(retType, params));
+        return method(refType, name, functionType(retType, params));
     }
 
     static MethodDesc method(TypeElement refType, String name, TypeElement retType, List<? extends TypeElement> params) {
-        return method(refType, name, methodType(retType, params));
+        return method(refType, name, functionType(retType, params));
     }
 
     // Copied code in jdk.compiler module throws UOE
     static MethodDesc ofString(String s) {
 /*__throw new UnsupportedOperationException();__*/        return java.lang.reflect.code.parser.impl.DescParser.parseMethodDesc(s);
     }
+
+
+    // MethodTypeDesc factories
+    // @@@ Where else to place them?
+
+    static FunctionType ofNominalDescriptor(MethodTypeDesc d) {
```

Comment:
These seems straightforward mapping from MTD -> FT (and back). So it feels like they should go in FunctionType.

Simiilarly, classDesc -> TE mapping (and maybe back) should be in either TypeElement or JavaType.

Reply:
Given `FunctionType` is not specific to Java types i was reluctant to place Java-specific functionality on it. The current location is  temporary. `JavaType` has some functionality to mediate between `ClassDesc`, may be that is the right place to place mediate with other nominal descriptors, or maybe we need a specific place for this, resolution, and reasoning about specific types?

I wanted to wait until we improve the modeling of Java types, and rethink the naming/location of the (Java-specific) `*Desc` classes.